tools/ui-components/src/button/button.tsx

```diff
@@ -58,27 +96,81 @@ const computeClassNames = ({
   return classNames.join(' ');
 };
 
-export const Button = React.forwardRef<HTMLButtonElement, ButtonProps>(
+export const Button = React.forwardRef<
+  HTMLButtonElement | HTMLAnchorElement,
+  ButtonProps
+>(
   (
     {
       variant = 'primary',
       size = 'medium',
       type = 'button',
       onClick,
-      children
+      children,
+      disabled,
+      block,
+      to,
+      target
     },
     ref
   ) => {
     const classes = useMemo(
-      () => computeClassNames({ size, variant }),
-      [size, variant]
+      () => computeClassNames({ size, variant, disabled, block }),
+      [size, variant, disabled, block]
     );
 
-    return (
-      <button ref={ref} className={classes} type={type} onClick={onClick}>
-        {children}
-      </button>
+    // Manually prevent click event if the button is disabled
+    // as `aria-disabled` marks the element disabled but still registers the click event.
+    // Ref: https://css-tricks.com/making-disabled-buttons-more-inclusive/#aa-the-difference-between-disabled-and-aria-disabled
+    const handleClick = useCallback(
+      (event: React.MouseEvent<HTMLButtonElement>) => {
+        const ariaDisabled = event.currentTarget.getAttribute('aria-disabled');
+
+        if (!ariaDisabled && onClick) {
+          onClick(event);
+        }
+      },
+      [onClick]
     );
+
+    const renderButton = useCallback(() => {
+      return (
+        <button
+          ref={ref as React.ForwardedRef<HTMLButtonElement>}
+          className={classes}
+          type={type}
+          onClick={handleClick}
+          aria-disabled={disabled}
+        >
+          {children}
+        </button>
+      );
+    }, [children, classes, ref, type, handleClick, disabled]);
+
+    const renderLink = useCallback(() => {
+      // Render a `button` tag if `disabled` is defined to keep the component semantically correct
+      // as a link cannot be disabled.
+      if (disabled) {
+        return renderButton();
+      }
+
+      return (
+        <a
+          ref={ref as React.ForwardedRef<HTMLAnchorElement>}
+          className={classes}
+          href={to}
+          target={target}
+        >
+          {children}
+        </a>
+      );
+    }, [children, classes, ref, disabled, to, target, renderButton]);
+
+    if (to) {
+      return renderLink();
+    } else {
+      return renderButton();
+    }
```

Comment:
The rendering logic here is a bit complicated in order to make our component semantically correct:
- If the `to` prop is defined, we render an `a` element
- If both `to` and `disabled` are defined, we render a `button` element because a link cannot be disabled
- If the `to` prop is not defined, we render a `button` element

Reply:
As a matter of curiosity, when would there be both a `to` and `disabled`?

Reply:
Hopefully, never :), just in case we pass both by mistake the element would just return a button element.

Reply:
Yup. This is me being paranoid. There should never be a case where both `to` and `disabled` are passed in.

I believe we can enforce this rule with TypeScript using union types, but the typing ended up a bit crazy in my initial attempt, so I decided to just keep the interface as-is and silently handle invalid use case for now.